### PR TITLE
Setup `duecredit` for machine-readable in-code citations

### PR DESCRIPTION
**Pull request by janosh**
Following this morning's MP Foundation meeting where I took this over from @mkhorton, this is a draft PR that adds `duecredit` class decorators and module citations to a dozen files to gather feedback before doing the rest.

@mkhorton Let me know if this is what you had in mind and I'll do the rest.

### CODE REVIEW

**Comment by mkhorton** [comment]
Yep, this looks perfect -- thanks @janosh.

**Comment by janosh** [comment]
Yay, took a bit longer than expected but I think I did added `duecredit` decorators for all DOIs in doc strings. Have to admit my focus was wavering towards the end so might have missed a few. But we can always add more later.

**Comment by JaGeo** [comment]
Don't worry! Well done finding so many!

**Comment by shyuep** [comment]
I think this proves my point that the onus shouldn't be on Mp dev to do this. The contributor can do this themselves. Seriously why waste the time?

**Comment by shyuep** [comment]
Anyway, if you want to ignore my advice in time management and prioritization, knock yourselves out doing unimportant stuff like this. It is not my time that is being wasted.

**Comment by utf** [comment]
I have one concern about visibility of the references. Many users will find pymatgen functionality in the API docs (this is often the page returned by google searches rather than the GitHub source). Currently the paper references are included in the docstrings themselves, so you can immediately see the source for a certain feature. E.g. see the docstring for some of the [dimensionality functions](https://pymatgen.org/pymatgen.analysis.dimensionality.html#pymatgen.analysis.dimensionality.get_dimensionality_larsen).

These references are being remove in favour of duecredit decorators but is there a way to automatically include the references in the documentation built by sphinx?

**Comment by mkhorton** [comment]
Good point @utf, I guess the decorator could patch the `__doc__` string? Could be a nice way to standardise references in the docstring too.

I also saw someone started something to collect them all in one place: https://github.com/duecredit/duecredit/pull/183/files

**Comment by mkhorton** [comment]
If it's helpful at all, I can a snippet:

```
from duecredit import Doi
def format_doi(self):
    """
    Processes a DOI into a formatted reference via Crossref API.
    (Alternatively could cache these somewhere.)
    
    Returns a Sphinx-compatible string with URL, intended to be 
    used with Sphinx documentation generation only.
    """
    
    from habanero import cn
    # choose an appropriate style here, slice is to cut off newline
    ref = cn.content_negotiation(ids=self.doi, format="text", style="apa")[:-2]
    
    return f"`{ref} <https://doi.org/{self.doi}>`_"
Doi.format_doi = format_doi
```

I imagine you can then wrap the original `cite`/`dcite` with functools.wrap, and modify the `__doc__` in place.

File: pymatgen/analysis/chemenv/coordination_environments/tests/test_coordination_geometry_finder.py
Comment (JaGeo) [inline pymatgen/analysis/chemenv/coordination_environments/tests/test_coordination_geometry_finder.py:1]: My fault. Removed the printing option some time ago

**Comment by janosh** [comment]
> Anyway, if you want to ignore my advice in time management and prioritization, knock yourselves out doing unimportant stuff like this.

@shyuep We're not going to town with this. My total time spent on this yesterday was ~2.5 hours incl. the MP update call where we discussed reviving this effort.

> This is something that can and should be crowd sourced to contributors.

@shyuep I disagree that external contributors should be asked to setup a systematic way for code citations. That's up to us to establish. Once we have that, we can ask contributors to follow our framework.

> https://github.com/materialsproject/foundation/pull/7#issuecomment-1608837201
> From Andrew: That said, how will users know about it and feel encouraged to use it? Presumably there will need to be something added to the docs in a visible way.

@arosen93 I'll add a checkbox to the PR template asking contributors to consider whether their code should one or more citations

**Comment by shyuep** [comment]
@janosh a systematic way involves writing in the contributor guidelines "Feel free to add die credit decorators for your contributed code, as well as a reference in the doc."  It does not involve spending time going through existing code to do it ourselves. 

Adding due credit for something like Ewald summation to the ewald code is not what this is meant for. Why not add duecredit to the VASP publication for the entire vasp package? Or for that matter, due credit to Kohn and Sham for everything DFT related? The Ewald class was written mostly by ME, with contributions from a few others over the years. If anything, I should be given due credit for that module. I just don't give enough of a shit about claiming credit for it.

**Comment by janosh** [comment]
> If anything, I should be given due credit for that module.

You _are_ getting credit for that! It's all recorded in `git` history which I care more about than publications tbh when I want to know who did the actual work.

Anyway, I take your point that setting up a citation system is not 1st priority and so shouldn't become a major time drain (which it wasn't).

**Pull request merged**